flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/BoundedProcessingOverRowProcessFunction.scala

```diff
@@ -0,0 +1,182 @@
+  private var output: Row = _
+  private var counterState: ValueState[Long] = _
+  private var counter : Long = _
+  private var smallestTsState: ValueState[Long] = _
+  private var smallestTs : Long = _
+  
+  override def open(config: Configuration) {
+    
+    output = new Row(forwardedFieldCount + aggregates.length)
+    // We keep the elements received in a list state 
+    // together with the ingestion time in the operator
+    // we also keep counter of processed elements
+    // and timestamp of oldest element
+    val rowListTypeInfo: TypeInformation[JList[Row]] =
+      new ListTypeInfo[Row](inputType).asInstanceOf[TypeInformation[JList[Row]]]
+    
+    val mapStateDescriptor: MapStateDescriptor[Long, JList[Row]] =
+      new MapStateDescriptor[Long, JList[Row]]("windowBufferMapState",
+        BasicTypeInfo.LONG_TYPE_INFO.asInstanceOf[TypeInformation[Long]], rowListTypeInfo)
+    
+    rowMapState = getRuntimeContext.getMapState(mapStateDescriptor)
+
+    val stateDescriptor: ValueStateDescriptor[Row] =
+      new ValueStateDescriptor[Row]("aggregationState", aggregatesTypeInfo)
+      
+    accumulatorState = getRuntimeContext.getState(stateDescriptor)
+    
+    val processedCountDescriptor : ValueStateDescriptor[Long] =
+       new ValueStateDescriptor[Long]("processedCountState", classOf[Long])
+    
+    counterState = getRuntimeContext.getState(processedCountDescriptor)
+    
+    val smallesTimestampDescriptor : ValueStateDescriptor[Long] =
+       new ValueStateDescriptor[Long]("smallesTSState", classOf[Long])
+    
+    smallestTsState = getRuntimeContext.getState(smallesTimestampDescriptor)
+    
+  }
+
+  override def processElement(
+    input: Row,
+    ctx: ProcessFunction[Row, Row]#Context,
+    out: Collector[Row]): Unit = {
+    
+    val currentTime = ctx.timerService().currentProcessingTime()
+    var i = 0
+
+    accumulators = accumulatorState.value()
+    // initialize state for the first processed element
+    if(accumulators == null){
+      accumulators = new Row(aggregates.length)
+      while (i < aggregates.length) {
+        accumulators.setField(i, aggregates(i).createAccumulator())
+        i += 1
+      }
+    }
+    
+    // get smallest timestamp 
+    smallestTs = smallestTsState.value()
+    if(smallestTs == 0L){
+      smallestTs = currentTime
+    }
+    // get previous counter value
+    counter = counterState.value()
+    
+    if (counter == precedingOffset) {
+      val retractTs = smallestTs
+      val retractList = rowMapState.get(smallestTs)
+
+      // get oldest element beyond buffer size   
+      // and if oldest element exist, retract value
+      i = 0
+      while (i < aggregates.length) {
+        val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+        aggregates(i).retract(accumulator, retractList.get(0).getField(aggFields(i)))
+        i += 1
+      }
+      retractList.remove(0)
+      counter -= 1
```

Comment:
Once `counter` is equal to `precedingOffset` we always have to accumulate and retract. So we only have to increment it as long as it is smaller. This can be done in an `else` branch of the retraction condition (`counter == precedingOffset`). Also we only need to update the `counterState` if we increment the counter.